### ISSUE TITLE
Character separation issue when typing in Korean

### ISSUE THREAD

**Issue by ycy3723** [neovide/neovide#2014]
Korean is a language where consonants and vowels combine to form a single syllable.

When changing the OS input source to Korean and typing in input mode, the first syllable entered is separated.
(Immediately after changing the input source, only the first character is separated, and subsequent typing is entered correctly.)

**Screenshots**
<img width="541" alt="image" src="https://github.com/neovide/neovide/assets/38966985/2c229681-8b5d-48c9-98c8-ca440e57941e">

https://github.com/neovide/neovide/assets/38966985/98cdfdf9-8b95-4bf2-bf0e-e6fd853381d5

 - OS: macOS 13.2.1
 - Neovide Version: 0.11.1
 - Neovim Version: 0.9.1

**Comment by fredizzimo** [comment]
Can you type that as plain text as well instead of just images, so that I can copy and paste it to Neovide? That's a lot easier for me than to try to type it in manually, especially without a Korean keyboard.

**Comment by ycy3723** [comment]
@fredizzimo Thank you for your assistance.

Ex1)
typing: ㅇㅏㄴㄴㅕㅇㅎㅏㅅㅔㅇㅛ
expected: 안녕하세요
neovide: ㅇㅏㄴ녕하세요

Ex2)
typing: ㄴㅔㅇㅗㅂㅣㄷㅡ
expected: 네오비드
neovide: ㄴㅔ오비드

**Comment by ycy3723** [comment]
Sorry my bad. I miss clicked.

**Comment by fredizzimo** [comment]
This might unfortunately be a macOS specific issue, which I can't help with myself.

I tried it on Windows, first by copypasting the expected text, since I was assuming that it was some kind of rendering problem. But that one worked fine.

Then I installed the Korean keyboard in Windows. And entered the text using the Windows IME, and that was also rendered correctly. 

BTW, the first example is the `dkssudktpdy` qwerty keys if someone needs to type it and the second one `spdhqiem`.

One thing I observed is that there appears to be an extra space when you type it the second time, it's not aligned with the rest of the column. That does not happen for me either.

**Comment by fredizzimo** [comment]
@ycy3723, I think that someone on macOS needs to debug the issue and try to find out exactly what's going on.

But there's one thing you could try, could you try the Alacritty terminal emulator? It uses the same Winit library for IME input as we do. If the issue occurs there, then the bug should be reported here https://github.com/rust-windowing/winit

**Comment by ycy3723** [comment]
@fredizzimo I tested now. But Alacritty is all good.

https://github.com/neovide/neovide/assets/38966985/2988e6b8-6c9a-40e7-826f-93b2dd0c644c

**Comment by fredizzimo** [comment]
@ycy3723, the log might contain some useful information

If you run `neovide --log`, then a file `neovide_rCurrent.log` should be created in the current directory. 

Note that this file gets quite big, and contains everything you type and see in Neovim, so don't do anything else than start Neovim, type the sequence and exit. That ensures that the file stays reasonably sized and the logging from the start is still included.

**Comment by ycy3723** [comment]
@fredizzimo Alacritty has a same issue. Here is `neovide_rCurrent.log`

[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12575440/neovide_rCURRENT.log)

**Comment by ycy3723** [comment]
[https://github.com/alacritty/alacritty/issues/936](https://github.com/alacritty/alacritty/issues/936)

Same issue on Alacritty.

**Comment by fredizzimo** [comment]
Do you think you can report a bug in https://github.com/rust-windowing/winit? 

According to the logs this is what happens.
1. First a regular keyboard input with ㅇ
2. IME commit with ㅏ
3. IME commit ㄴ
4. IME commit 녕
5. IME commit 녕
6. IME commit 하
7. IME commit 세
8. IME commit 요

The issue is probably that the first ㅇcomes as a regular keyboard input and is removed from the IME system. Then when ㅏand ㄴare typed, they are not valid sequences, so it forces them as individual glyphs. This is just pure speculation, but it could be that the IME system is enabled too late or something.

**Comment by ycy3723** [comment]
@fredizzimo  Your speculation seems very reasonable. It has been very helpful. I will indeed report the bug as you suggested. Thank you once again 💯💯💯

**Comment by leokang94** [comment]
@fredizzimo hello!

I used neovide so attractive that I use it well despite the inconvenience of Korean.

If I don't know much about IME and just report the phenomenon, I have the following problem, is it the same reason as this issue?

> // first, dot is not typed.
> typing: ㅇㅏㄴㄴㅕㅇㅎㅏㅅㅔㅇㅛ.
> expected: 안녕하세요.
> neovide: 안녕하세요 // dot is now showed.
>
> // second, space is doubled.
> typing: ㅂㅏㄴㄱㅏㅂㅅㅡㅂㄴㅣㄷㅏ.\<space\>
> expected: 반갑습니다.\<space\>
> neovide: 반갑습니다\<space\>\<space\>
>
> // Not in the video... but even when I did this!
> typing: ㅂㅏㄴㄱㅏㅂㅅㅡㅂㄴㅣㄷㅏ\<right-arrow\>\<space\>
> expected: 반갑습니다\<space\>
> neovide: 반갑습니다\<space\>\<space\>

https://github.com/user-attachments/assets/0a7b96bf-6d7f-4586-9c2c-c0125d6185c2

If I type korean and then swap to english, it works fine from then on. There seems to be an issue with characters like !, ?, ., etc. when in korean IME.

When I tested the same case in wezterm, there's no problem (in particular, wezterm shows the consonants/vowels in korean merging together!)

https://github.com/user-attachments/assets/a2c8cf07-91d8-4300-a30c-aecdca237b2d

(oops, in this video, I typed wrong word! `find` -> `fine` 😅)

**Comment by fredizzimo** [comment]
Yes, I think your issue is related, and also maybe related to this:
* https://github.com/neovide/neovide/issues/2567

Unfortunately, there's not much we can do on the Neovide side, since the bug appears to be in Winit. I also don't have a macOS machine, so I'm not able to debug and figure out what's going on there. 

The Winit maintainers are also busy with other stuff, especially the macOS, ones, so if someone is able to feel free to help them out.

**Comment by leokang94** [comment]
@fredizzimo  Thank you! I thought it might be related to that issue, and it turns out you're right.
I don't know much about rust and how Winit works, but I'll keep an eye on that issue for now. (Actually, I was aware of the Korean issue and was using it anyway, lol)
